include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel

```diff
@@ -0,0 +1,587 @@
+                    auto const idx = blockBeginIdx + idxInSuperCell;
+                    // with the current Yee grid, no shift needed here
+                    auto const pmlIdx = precisionCast< float_X >( idx );
+                    auto const coeffs = detail::getCoefficients(
+                        pmlIdx,
+                        parameters,
+                        dt
+                    );
+
+                    if( detail::isInPML( coeffs ) )
+                    {
+                        /* This precomputation of partial derivatives is done
+                         * more for readability, rather than avoiding computing
+                         * it twice
+                         */
+                        using Difference = typename T_Curl::Difference;
+                        const typename Difference::template GetDifference< 0 > Dx;
+                        const typename Difference::template GetDifference< 1 > Dy;
+                        const typename Difference::template GetDifference< 2 > Dz;
+                        auto const localB = cachedB.shift( idxInSuperCell );
+                        auto const dBxDy = Dy( localB ).x( );
+                        auto const dBxDz = Dz( localB ).x( );
+                        auto const dByDx = Dx( localB ).y( );
+                        auto const dByDz = Dz( localB ).y( );
+                        auto const dBzDx = Dx( localB ).z( );
+                        auto const dBzDy = Dy( localB ).z( );
+
+                        /* Update convolutional fields using [Taflove, Hagness],
+                         * eq. (7.105a,b) and similar for other components.
+                         * For PIC the right-hand side uses B, not H.
+                         */
+                        auto & psiE = fieldPsiE( idx );
+                        psiE.yx = coeffs.b.x( ) * psiE.yx + coeffs.c.x( ) * dBzDx;
+                        psiE.zx = coeffs.b.x( ) * psiE.zx + coeffs.c.x( ) * dByDx;
+                        psiE.xy = coeffs.b.y( ) * psiE.xy + coeffs.c.y( ) * dBzDy;
+                        psiE.zy = coeffs.b.y( ) * psiE.zy + coeffs.c.y( ) * dBxDy;
+                        psiE.xz = coeffs.b.z( ) * psiE.xz + coeffs.c.z( ) * dByDz;
+                        psiE.yz = coeffs.b.z( ) * psiE.yz + coeffs.c.z( ) * dBxDz;
+
+                        /* [Taflove, Hagness], eq. (7.106) and similar for other
+                         * components. Coefficients Ca, Cb as given in (7.107a,b)
+                         * are general to account for materials, in addition to
+                         * artificial PML absorbing medium. We do not have any
+                         * real material, so in (7.107a,b) have to use
+                         * sigma(i + 1/2, j, k) = 0 (it is another sigma,
+                         * unrelated to PML), eps(i + 1/2, j, k) = EPS0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
+                         * not H, in the right-hand side.
```

Comment:
~in~ on [see here](https://www.dict.cc/englisch-deutsch/on+the+right+hand+side.html)

Reply:
Good point.